samples/can/src/main.c

```diff
@@ -0,0 +1,336 @@
+	}
+}
+
+void tx_thread(void *can_dev_param, void *unused2, void *unused3)
+{
+	struct can_msg msg;
+	struct can_msg msg_button_cnt;
+	struct device *can_dev = can_dev_param;
+	static u8_t counter;
+	static u16_t button_press_cnt;
+
+	msg.std_id = LED_MSG_ID;
+	msg.id_type = CAN_STANDARD_IDENTIFIER;
+	msg.dlc = 1;
+	msg.rtr = CAN_DATAFRAME;
+	msg.data[0] = 0;
+
+	msg_button_cnt.std_id = BUTTON_MSG_ID;
+	msg_button_cnt.id_type = CAN_STANDARD_IDENTIFIER;
+	msg_button_cnt.dlc = 2;
+	msg_button_cnt.rtr = CAN_DATAFRAME;
+	msg_button_cnt.data[0] = 0;
+	msg_button_cnt.data[1] = 0;
+
+	printk("TX thread is running.\n");
+	while (1) {
+		k_sem_take(&tx_sem, K_FOREVER);
+		button_press_cnt++;
+		counter = counter >= NUM_LEDS ? 0 : counter + 1;
+		printk("Button pressed! Send message %u\n", counter);
+		msg.data[0] = counter;
+		msg_button_cnt.data[0] = button_press_cnt & 0xFF;
+		msg_button_cnt.data[1] = (button_press_cnt >> 8) & 0xFF;
+		can_send(can_dev, &msg, 10, tx_irq_callback);
+		can_send(can_dev, &msg_button_cnt, 10, NULL);
+		if (counter == NUM_LEDS) {
+			send_string("Counter is at "NUM_LEDS_STR". Reset\n", can_dev);
+		}
+	}
+}
+
+void rx_str_thread(void *msgq, void *can_dev_param, void *unused)
+{
+	struct can_msg msg;
+
+	{
+		struct can_filter filter = {
+			.id_type = CAN_EXTENDED_IDENTIFIER,
+			.rtr = CAN_DATAFRAME,
+			.ext_id = STR_MSG_ID,
+			.rtr_mask = 1,
+			.ext_id_mask = CAN_EXT_ID_MASK
+		};
+		struct device *can_dev = can_dev_param;
+
+		can_attach_msgq(can_dev, msgq, &filter);
+	}
+	while (1) {
+		k_msgq_get((struct k_msgq *)msgq, &msg, K_FOREVER);
+		for (int i = 0; i < msg.dlc; i++)
+			printk("%c", msg.data[i]);
+	}
+}
+
+void led_thread(void *msgq, void *can_dev_param, void *gpio_dev_param)
+{
+	struct can_msg msg;
+	struct device *gpio_dev = gpio_dev_param;
+	int ret;
+	int filter_id;
+
+	ret = gpio_pin_configure(gpio_dev, CONFIG_PIN_LED_1, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_dev, CONFIG_PIN_LED_1, 0);
+	#ifdef CONFIG_USE_LED_2
+	ret |= gpio_pin_configure(gpio_dev, CONFIG_PIN_LED_2, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_dev, CONFIG_PIN_LED_2, 0);
+	#endif
+	#ifdef CONFIG_USE_LED_3
+	ret |= gpio_pin_configure(gpio_dev, CONFIG_PIN_LED_3, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_dev, CONFIG_PIN_LED_3, 0);
+	#endif
+	#ifdef CONFIG_USE_LED_4
+	ret |= gpio_pin_configure(gpio_dev, CONFIG_PIN_LED_4, GPIO_DIR_OUT);
+	gpio_pin_write(gpio_dev, CONFIG_PIN_LED_4, 0);
+	#endif
+
+	if (ret) {
+		printk("ERROR configure pins\n");
+		return;
+	}
+
+	{
+		struct can_filter filter = {
+			.id_type = CAN_STANDARD_IDENTIFIER,
+			.rtr = CAN_DATAFRAME,
+			.std_id = LED_MSG_ID,
+			.rtr_mask = 1,
+			.std_id_mask = CAN_STD_ID_MASK
+		};
+
+		struct device *can_dev = can_dev_param;
+
+
+		filter_id = can_attach_msgq(can_dev, msgq, &filter);
+		printk("filter id: %d\n", filter_id);
+	}
+	while (1) {
+		k_msgq_get((struct k_msgq *)msgq, &msg, K_FOREVER);
+		if (msg.dlc != 1) {
+			continue;
+		}
+		switch (msg.data[0]) {
+		case 0:
```

Comment:
Could you use macros that would self-describe the case?

Reply:
Sure.